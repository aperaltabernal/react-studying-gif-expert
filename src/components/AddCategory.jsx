import { useState } from "react";
import PropTypes from 'prop-types';

export const AddCategory = ({onNewCategory}) => {
    const [inputValue, setInputValue] = useState('');

    const onInputChange = ({target}) => {
        setInputValue(target.value);
    }

    const onSubmit = (event) => {
        console.log("onSubmit ejecutado");
        event.preventDefault();//Evita que la página se refresque por ejecutar el onSubmit del form
        console.log("inputValue: " + inputValue)
        if(inputValue.trim().length <= 1) return;

        //El componenten sólo debe encargarse de emitir el valor, no validar nada mas
        onNewCategory(inputValue);
        setInputValue('');
    }

    return (
        <form onSubmit={onSubmit} aria-label="form">{/* Al presionar enter, el elemento form recarga toda la pagina */}
            <input
                type="text"
                placeholder="Buscar gifs"
                value={inputValue}
                onChange={onInputChange}
            />
        </form>
    );
}

AddCategory.propTypes = {
    onNewCategory: PropTypes.func.isRequired
}