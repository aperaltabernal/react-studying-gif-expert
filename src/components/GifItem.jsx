import PropTypes from 'prop-types';

export const GifItem = ({title, url, id}) => {
    return (
        <div className="card">
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
}

//npm install --save prop-types
/*Las siguientes propTypes hacen que el campo title y url sean obligatorios
*/
GifItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}