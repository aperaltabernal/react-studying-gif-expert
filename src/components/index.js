/* Este es un archivo de barril
en el cual se pueden centralizar exportaciones
de tal forma que se llama el nombre de la carpeta donde se encuentre
y ya es suficiente para usarlo.

Por ejemplo:

    import {AddCategory, GifGrid} from './components' */

export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';