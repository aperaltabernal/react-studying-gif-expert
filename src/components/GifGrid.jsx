import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifItem } from './GifItem';

export const GifGrid = ({category}) => {
    
    const {images, isLoading} = useFetchGifs(category);

    console.log(`Rendering ${category}`);
    console.log(isLoading)

    return (
        <>
            <h3>{category}</h3>
            {
                isLoading && (<h2>Cargando...</h2>)
            }
            <div className='card-grid'>
                {/*imagenes.map(({id, title}) => <li key={id}>{title}</li>)*/}
                {   
                    images.map((image) => (
                            <GifItem 
                                key={image.id}
                                { ...image }//Tecnica=Esparsir las propiedades, con esto el componente recibirá en las props los elementos hijos del objeto que se está pasando.
                            />
                        )
                    )
                }
            </div>
        </>
    );
}