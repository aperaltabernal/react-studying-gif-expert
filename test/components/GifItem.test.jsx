import { render, screen } from '@testing-library/react';
import { GifItem } from '../../src/components/GifItem';

describe('Pruebas en <GifItem />', () => { 
    const title = 'Read';
    const url = 'http://reading.com/read.jpg';

     test('Debería realizar match con el snapshot', () => { 
        const {container} = render(<GifItem url={ url } title={ title }/>);
        expect(container).toMatchSnapshot();
      });

    test('Debería mostrar la imagen con el URL y el ALT indicado', () => {
        render(<GifItem title={title} url={url} />);
        screen.debug();
        const {src, alt} = screen.getByRole('img');
        expect(src).toBe(url);
        expect(alt).toBe(title);
     });

     test('Debería mostrar el titulo en el componente', () => { 
        render(<GifItem title={title} url={url}/>);
        expect(screen.getByText(title)).toBeTruthy();
      });
 });