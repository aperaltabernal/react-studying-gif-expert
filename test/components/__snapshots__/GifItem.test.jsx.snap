// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem /> Deber√≠a realizar match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Read"
      src="http://reading.com/read.jpg"
    />
    <p>
      Read
    </p>
  </div>
</div>
`;
